### PR TITLE
[docs][scripts] fix broken README.md weblinks and refactor protoc-generator.sh

### PR DESCRIPTION
**Pull request by axonasif**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,24 +33,24 @@ Tightly integrated with GitLab, GitHub, and Bitbucket, Gitpod automatically and
...
-Gitpod is provided as a [managed Saas version](https://gitpod.io) with a free subscription for open-source or a [free self-hosted version](https://www.gitpod.io/self-hosted/). An enterprise license is available [here](https://www.gitpod.io/self-hosted/).
+Gitpod is provided as a [managed Saas version](https://gitpod.io) with a free subscription for open-source or a [free self-hosted version](https://www.gitpod.io/self-hosted). An enterprise license is available [here](https://www.gitpod.io/self-hosted).
 
 ## Getting Started
 
 You can start using Gitpod with one or more of the following ways:
-1. [Use a Prefixed URL](https://www.gitpod.io/docs/context-urls/)
-1. [Install Browser Extension](https://www.gitpod.io/docs/browser-extension/)
-1. [Enable GitLab Integration](https://www.gitpod.io/docs/gitlab-integration/)
+1. [Use a Prefixed URL](https://www.gitpod.io/docs/getting-started/#prefixed-url)
```
Comment (jankeromnes) [inline README.md:41]: **Praise**: Nice touch! (For me, the old link still works, but the one you chose works better here 👍)

File: README.md [hunk 2]
```diff
@@ -81,4 +81,4 @@ To this end we've developed a number of open source projects including:
 ## Code of Conduct
 
 We want to create a welcoming environment for everyone who is interested in contributing to Gitpod or participating in discussions with the Gitpod community.
-This project has adopted the [Contributor Covenant Code of Conduct](https://github.com/gitpod-io/gitpod/blob/master/CODE_OF_CONDUCT.md), [version 2.0](https://www.contributor-covenant.org/version/2/0/code_of_conduct/).
+This project has adopted the [Contributor Covenant Code of Conduct](https://github.com/gitpod-io/gitpod/blob/master/CODE_OF_CONDUCT.md), [version 2.0](https://www.contributor-covenant.org/version/2/0/code_of_conduct).
```
Comment (jankeromnes) [inline README.md:84]: **Minor**: The canonical link here still seems to have a final `/`.

```suggestion
This project has adopted the [Contributor Covenant Code of Conduct](https://github.com/gitpod-io/gitpod/blob/master/CODE_OF_CONDUCT.md), [version 2.0](https://www.contributor-covenant.org/version/2/0/code_of_conduct/).
```

File: README.md [hunk 3]
```diff
@@ -33,24 +33,24 @@ Tightly integrated with GitLab, GitHub, and Bitbucket, Gitpod automatically and
 
 [Learn more 👉](https://www.gitpod.io/features/)
 
-Gitpod is provided as a [managed Saas version](https://gitpod.io) with a free subscription for open-source or a [free self-hosted version](https://www.gitpod.io/self-hosted/). An enterprise license is available [here](https://www.gitpod.io/self-hosted/).
+Gitpod is provided as a [managed Saas version](https://gitpod.io) with a free subscription for open-source or a [free self-hosted version](https://www.gitpod.io/self-hosted). An enterprise license is available [here](https://www.gitpod.io/self-hosted).
```
Comment (jankeromnes) [inline README.md:36]: **Minor**: The canonical link here still seems to have a final `/`:

```suggestion
Gitpod is provided as a [managed Saas version](https://gitpod.io) with a free subscription for open-source or a [free self-hosted version](https://www.gitpod.io/self-hosted/). An enterprise license is available [here](https://www.gitpod.io/self-hosted/).
```

**Review by jankeromnes — COMMENTED** [review]
Hey, many thanks for your contribution @axonasif!

The README.md link fixes look good to me, but I'm unsure about the `protoc-generator.sh` refactor.
- Could you please explain why the refactor was necessary?
- It might be worth separating these two changes (README.md fixes & protoc refactor) in two separate Pull Requests

**Comment by jankeromnes** [comment]
Requesting review from @aledbf for the `protoc-generator.sh` refactor.

File: README.md (hunk 1 above)
Comment (axonasif) [reply README.md:41]: Feels weird now, yesterday night few of the links were resulting in `404` pages on the gitpod.io side for me, specifically for the getting-started links in `README.md`. Now all of them seem to be working :no_mouth: Sorry about it.

File: README.md (hunk 3 above)
Comment (axonasif) [reply README.md:36]: Got it.

**Comment by aledbf** [comment]
@axonasif the `typescript_protoc` function change makes sense but not the `install_dependencies`. What are you trying to change?

**Comment by axonasif** [comment]
> The README.md link fixes look good to me, but I'm unsure about the `protoc-generator.sh` refactor.
> 
>     * Could you please explain why the refactor was necessary?

As the title says, the refactoring inside `install_dependencies()` function is meant to simplify it.
And on line-31(initial) of [here](https://github.com/gitpod-io/gitpod/pull/4027/commits/78f0ee896cfff5583a7324787fea53b6defaa2ec#diff-539ce15be6133bf4370ed98354219448db52be720348eb600519a032ba340b0cR45), `pwd` could potentially throw an error because of assigning a value after the `local` definition but not in a separate instance even though `shopt -o errexit` is being set from the caller scripts (e.g [generate.sh](https://github.com/gitpod-io/gitpod/blob/d114f037a1c34df5206bb93fc9fe48481f28bfd7/components/registry-facade-api/generate.sh#L7))

>     * It might be worth separating these two changes (README.md fixes & protoc refactor) in two separate Pull Requests

Will keep that in mind @jankeromnes  :smile:

**Comment by axonasif** [comment]
> @axonasif the `typescript_protoc` function change makes sense but not the `install_dependencies`. What are you trying to change?

Just to simplify the process. In my view using loops instead of repeated commands for the same kind of task seems better.

**Comment by aledbf** [comment]
> In my view using loops instead of repeated commands for the same kind of task seems better.

Not in my personal opinion. Please rollback the change in `install_dependencies`.

**Comment by axonasif** [comment]
> > In my view using loops instead of repeated commands for the same kind of task seems better.
> 
> Not in my personal opinion. Please rollback the change in `install_dependencies`.

Okay, understood.

**Comment by aledbf** [comment]
@axonasif please squash the commits and this LGTM

**Comment by axonasif** [comment]
> @axonasif please squash the commits and this LGTM

I'm on it, hold on a bit please.

**Comment by axonasif** [comment]
> @axonasif please squash the commits and this LGTM

It's done a while back BTW

**Review by aledbf — APPROVED** [review]
LGTM

**Review by jankeromnes — APPROVED** [review]
Looks good to me too, thanks!

**Pull request merged**